### PR TITLE
Array refine

### PR DESCRIPTION
**Pull request by odashi**
This change involves some broad changes about `Array`'s interface:

- Rename some functions to  consolidate naming of getter/setters.
- Add functionality to get/set all values.
- Add fill function.
- Changes some return types.
- Add detailed tests.

### CODE REVIEW

File: src/buffer.rs [hunk 1]
```diff
@@ -71,16 +71,31 @@ impl<'hw> Buffer<'hw> {
...
     /// Returns the const handle owned by this buffer.
     ///
     /// # Returns
     ///
     /// Owned handle as a const pointer.
-    pub(crate) unsafe fn as_handle(&self) -> *const u8 {
+    ///
+    /// # Safety
+    ///
+    /// This function returns a raw pointer of the inner memory or a haandle of the associated
```
Comment (tetsuok) [inline src/buffer.rs]:
```suggestion
    /// This function returns a raw pointer of the inner memory or a handle of the associated
```

File: src/buffer.rs [hunk 2]
```diff
@@ -71,16 +71,31 @@ impl<'hw> Buffer<'hw> {
...
+    /// hardware-specific object.
+    /// The value can not be used without knowing the associated hardware.
```
Comment (tetsuok) [inline src/buffer.rs]: I think the returned value would be clearer.
```suggestion
    /// The returned value can not be used without knowing the associated hardware.
```

File: src/buffer.rs [hunk 3]
```diff
@@ -89,7 +104,13 @@ impl<'hw> Buffer<'hw> {
     /// # Returns
     ///
     /// Owned handle as a mutable pointer.
-    pub(crate) unsafe fn as_mut_handle(&mut self) -> *mut u8 {
+    ///
+    /// # Safety
+    ///
+    /// This function returns a raw pointer of the inner memory or a haandle of the associated
```
Comment (tetsuok) [inline src/buffer.rs]:
```suggestion
    /// This function returns a raw pointer of the inner memory or a handle of the associated
```

File: src/buffer.rs [hunk 4]
```diff
@@ -89,7 +104,13 @@ impl<'hw> Buffer<'hw> {
...
+    /// device-specific object.
+    /// The value can not be used without knowing the associated hardware.
```
Comment (tetsuok) [inline src/buffer.rs]:
```suggestion
    /// The returned value can not be used without knowing the associated hardware.
```

File: src/array.rs [hunk 5]
```diff
@@ -86,57 +87,169 @@ impl<'hw> Array<'hw> {
...
+        array
+    }
+
+    /// Creates a new `Array` with arbitrary shape on the specific hardware.
+    ///
+    /// # Arguments
+    ///
+    /// * `hardware`: `Hardware` object to host the value.
+    /// * `shape`: `Shape` of the new `Array`.
+    /// * `value`: Values copied to the underlying buffer. The number of values must be the same as
```
Comment (tetsuok) [inline src/array.rs]: nit: I was a bit confused about when `values` are copied to the buffer.

```suggestion
    /// * `values`: Values to be copied to the underlying buffer. The number of values must be the same as
```

File: src/array.rs [hunk 6]
```diff
@@ -86,57 +87,169 @@ impl<'hw> Array<'hw> {
...
+        array.set_values_f32(values)?;
+        Ok(array)
+    }
+
+    /// Creates a new `Array` with a specified `Shape` filled by a single value.
+    ///
+    /// # Arguments
+    ///
+    /// * `hardware`: `Hardware` object to host the value.
+    /// * `shape` - `Spahe` of the resulting `Array`.
```
Comment (tetsuok) [inline src/array.rs]:
```suggestion
    /// * `shape` - `Shape` of the resulting `Array`.
```

File: src/array.rs [hunk 7]
```diff
@@ -292,4 +405,515 @@ impl<'hw> Clone for Array<'hw> {
 }
 
 #[cfg(test)]
-mod tests {}
+mod tests {
```
Comment (tetsuok) [inline src/array.rs:408]: The tests have 500 lines (55% of this file). If the number of the tests grows in the near future, we might want to consider moving the tests to a dedicated file.

Comment (odashi) [reply src/array.rs:408]: Yes, but we may lack some advantages of visibility.

Comment (tetsuok) [reply src/array.rs:408]: That's exactly right.

**Review by tetsuok — APPROVED** [review]

**Pull request merged**